### DISCUSSION TITLE
What's the best approach for shared, auto-imported utilties between app and server?

### DISCUSSION THREAD

**Discussion by BracketJohn** [nuxt/nuxt#25708] [Questions]
Hey there!

As you know, you can use the `utils/` and `server/utils` respectively to have auto-imported, non-composable helpers for the app and server respectively.

We make use of this and like it, but are encountering an issue, that I'm sure others in the community also have: Some utilities (eg: `sleep` or `getFileExtension`) are required on both the server and the app. It would be great to declare them once and then use them in both places, but [as per the docs](https://nuxt.com/docs/guide/directory-structure/utils) these server- and app-utilities are not shared. We understand why this is the case, but would love to find a solution where some utilities that we know work & are safe for both app- and server to be shared & auto-imported.

We are looking for an experience share: Have you encountered this issue on your projects before? If so, did you find any way to improve / work-around this situation?

Thanks & have a great day 🧋

**Comment by AloisSeckar** [comment]
the easiest way is to create simple class just to re-export the thing declared on the other side, e.g.:

```
// /server/utils/serverUtils.ts

import { clientUtil } from '@/utils/clientUtil.ts`

export const serverUtil = clientUtil
```

Some extra piece of code, but no duplicated logic. And you decide which of your functions "work & are safe"

**Reply by BracketJohn** [reply]
Ah this looks nice & I especially love the fact that you explicitly declare what is safe and what not.

Thanks!

**Reply by BracketJohn** [reply]
I've come up with a different approach that is working better for us, see below: https://github.com/nuxt/nuxt/discussions/25708#discussioncomment-10543728

**Comment by Ed1ks** [comment]
unfortunately it doesn't take the types with it

**Reply by BracketJohn** [reply]
maybe my answer below will help?

see https://github.com/nuxt/nuxt/discussions/25708#discussioncomment-10543728

**Comment by BracketJohn** [comment]
I found a pattern which works better for our team:
1. create a root-level folder like `serverAndAppUtils`
2. add it as an auto-import dir of the server
    ```ts
    // in `defineNuxtConfig` of `nuxt.config.ts`
    nitro: {
      imports: {
        dirs: ['server/utils', 'serverAndAppUtils'],
      },
    }
    ```
3. add it as an auto-import dir of the app
    ```ts
    // in `defineNuxtConfig` of `nuxt.config.ts`
    imports: {
      dirs: ['composables', 'utils', 'serverAndAppUtils'],
    },
    ```


Now the utils defined inside `serverAndAppUtils` are available for both the server and the app, including types.

The name `serverAndAppUtils` is a bit unwieldy, but other than that it has been working great for our team:
- editor support
- types 
- clearly separated utils that will be used by both server and app (and thus cannot contain app or server-specific code)

**Reply by Ed1ks** [reply]
Thanks. I named it shared

**Comment by tag** [comment]
For those looking for an answer on newer versions, the "shared" folder exists by default.
https://nuxt.com/docs/4.x/api/nuxt-config#shared

**Reply by cco3** [reply]
But interestingly, it doesn't appear to be shared with itself. (That is, you do not get auto-imports in shared utils from shared utils)

**Reply by Ed1ks** [reply]
layers/layer/shared/types also dont get recognized from layer itself
